views.go

```diff
@@ -9,31 +9,107 @@ import (
 	"strings"
 )
 
-func Login(c echo.Context) error {
-	redir, err := url.Parse(c.FormValue("service"))
+func ValidateService(c echo.Context) *url.URL {
+	service, err := url.Parse(c.QueryParam("service"))
+	if err == nil {
+		return service
+		// fmt.Println(service.Host)
+		// switch service.Host {
+		// case "":
+		// 	fallthrough
+		// case "localhost:5000":
+		// 	fallthrough
+		// case "127.0.0.1:5000":
+		// 	fallthrough
+		// case "localhost:8000":
+		// 	fallthrough
+		// case "127.0.0.1:8000":
+		// 	fallthrough
+		// case "localhost:8080":
+		// 	fallthrough
+		// case "127.0.0.1:8080":
+		// 	return service
+		// }
+	}
+	return nil
+}
+
+func LoginPOST(c echo.Context) error {
+	data := new(TemplateData)
+
+	service := ValidateService(c)
+	if service == nil {
```

Comment:
`service` is used outside this `if`, will throw `undefined` error in this style, same for `username` and `verification_key`.